File: plugin/src/runtime/handshake.ts
import type { Capability, HelloAck, HelloRequest } from '../protocol/messages.js';
import { transitionStatus } from '../state/connection.js';
import type { ConnectionStatus } from '../state/connection.js';

export function createHelloRequest(
  logseqVersion: string,
  pluginVersion: string,
  capabilities: Capability[],
): HelloRequest {
  return {
    logseqVersion,
    pluginVersion,
    apiCapabilities: capabilities,
  };
}

export function handleHelloAck(
  status: ConnectionStatus,
  ack: HelloAck,
): { status: ConnectionStatus; enabledCapabilities: Capability[] } {
  const nextStatus = transitionStatus(status, 'ready');
  return {
    status: nextStatus,
    enabledCapabilities: ack.enabledCapabilities,
  };
}

File: mcp-server/src/runtime/handshake.rs
use crate::protocol::messages::{HelloAck, HelloRequest};
use crate::state::capabilities::CapabilityList;

pub fn handle_hello(request: HelloRequest, server_version: &str) -> HelloAck {
    let list = CapabilityList::from_capabilities(request.api_capabilities);
    HelloAck {
        server_version: server_version.to_string(),
        enabled_capabilities: list.enabled(),
    }
}
